### PR TITLE
Replace values columns

### PR DESCRIPTION
**Pull request by stevekogo**

### CODE REVIEW

File: instat/static/InstatObject/R/data_object_R6.R [hunk 1]
```diff
@@ -425,43 +425,98 @@ data_object$set("public", "remove_columns_in_data", function(cols=c()) {
...
   }
-  if(str_data_type == "numeric") {
-    new_value <- as.numeric(new_value)
+  for(col_name in col_names){
+    str_data_type <-self$get_variables_metadata(property = data_type_label, column = col_name)
+    if(str_data_type == "factor") {
+      if(!is.na(new_value) && !(new_value %in% levels(private$data[[col_name]]))) {
+        new_index <- which(levels(private$data[[col_name]]) == old_value)
+        levels(private$data[[col_name]])[new_index] <- new_value
+      }
```
Comment (dannyparsons) [inline instat/static/InstatObject/R/data_object_R6.R]: I think there's something not right here with factors. It seems all that is being done is if new_value is not an existing level then you are adding it as a level. But there is no replacement being done in the data. And if new_value is an existing level then nothing happens?
Also, I think for now we keep it strict and only allow replacement in factor columns with current levels (or NA). So if not we just give an error.

File: instat/static/InstatObject/R/data_object_R6.R [hunk 2]
```diff
@@ -425,43 +425,98 @@ data_object$set("public", "remove_columns_in_data", function(cols=c()) {
...
+    }
+    else{
+      if(!is.na(start_value) && !is.na(end_value)){
+        index <- which(private$data[[col_name]]>=start_value & private$data[[col_name]]<=end_value)
```
Comment (dannyparsons) [inline instat/static/InstatObject/R/data_object_R6.R]: This looks good. Can we add functionality here, and in the dialog for specifying open/closed end points. i.e. we should be able to do > or >= for start and end.
Only just thought of adding this, sorry.

File: instat/static/InstatObject/R/data_object_R6.R [hunk 3]
```diff
@@ -425,43 +425,98 @@ data_object$set("public", "remove_columns_in_data", function(cols=c()) {
...
+      else{
+        index <- which(private$data[[col_name]] == old_value)
+      }
+      if(str_data_type == "integer") {
+        #TODO Check that what checks are needed here
+        new_value <- as.integer(new_value)
+      }
+      if(str_data_type == "numeric") {
+        new_value <- as.numeric(new_value)
+      }
```
Comment (dannyparsons) [inline instat/static/InstatObject/R/data_object_R6.R]: There should also be a case for character columns.
And an else case to deal with all other types of columns.

File: instat/static/InstatObject/R/data_object_R6.R (hunk 1 above)
Comment (stevekogo) [inline instat/static/InstatObject/R/data_object_R6.R]: I don't understand your argument here but I have done some fixing which I still feel are inadequate. Have a look at the new pull request then we can have a discussion

**Comment by dannyparsons** [comment]
There's two points on factors:
1. I don't think we want to be changing factor levels through this method, since you should do that through the proper factor dialogs - and there you can rename levels. So if new_value is not an existing level then we should give an error and not add a new level to the factor.
2. I don't think anything would happen if I tried to do a new value in a factor column because you loop:
```
for (new_index in index){
    private$data[[col_name]][[new_index]] <- new_value
}
```
seems to be inside the Else, so it never runs for factors. Is that right, unless I'm reading it wrong?

**Comment by dannyparsons** [comment]
Is this ready to go in now?

**Comment by stevekogo** [comment]
I think so. Though I need to add warnings to the code but that can wait. 


Sent from Samsung Mobile

-------- Original message --------

**Pull request merged**